File: frontend/src/renderer/src/data/ProblemUpdate.ts
import { ProblemId } from '../services/ProblemService'

export interface ProblemUpdate {
  type: 'ongoing' | 'finished' | 'text'
  id: ProblemId
}

export interface FinishedProblemUpdate extends ProblemUpdate {
  successful: boolean
  solution: string | undefined
  error: string | undefined
}

export interface OngoingProblemUpdate extends ProblemUpdate {

}

export interface StartProblemUpdate extends OngoingProblemUpdate {

}

export interface TextProblemUpdate extends OngoingProblemUpdate {
  text: string
  lines: string[]
}

File: src/backend/Common/Updates/ProblemUpdate.cs
using System.Text.Json.Serialization;

namespace Common.Updates;

/// <summary>
/// Base type for updates emitted while running a problem.
/// </summary>
[JsonDerivedType(typeof(OngoingProblemUpdate))]
[JsonDerivedType(typeof(FinishedProblemUpdate))]
[JsonDerivedType(typeof(TextProblemUpdate))]
[JsonDerivedType(typeof(StartProblemUpdate))]
[JsonDerivedType(typeof(GlyphGridUpdate))]
[JsonDerivedType(typeof(StringGridUpdate))]
public abstract class ProblemUpdate
{
    /// <summary>
    /// Gets the discriminator used by clients to process the update.
    /// </summary>
    public abstract string Type { get; }

    /// <summary>
    /// Gets or sets the identifier for the problem that produced the update.
    /// </summary>
    public ProblemId Id { get; set; } = new();
}

